docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,450 @@
+
+Because the `Block` shared by IBC and Tendermint is produced and parsed using
+protobuf, new structures can be added to the Block without breaking the
+ability of legacy users to parse the new structure. Breaking changes between
+current users of IBC and new Tendermint blocks only occur if data that is
+relied upon by the current users is no longer included in the current fields.
+
+For the case of BLS aggregated signatures, a new `AggregatedSignature` field
+can therefore be added to the `Commit` field without breaking current users.
+Current users will be broken when counterparty chains upgrade to the new version
+and _begin using_ BLS signatures. Once counterparty chains begin using BLS
+signatures, the BlockID hashes will include hashes of the `AggregatedSignature`
+data structure that the legacy users will not be able to compute. Additionally,
+the legacy software will not be able to parse and verify the signatures to
+ensure that a supermajority of validators from the counterparty chain signed
+the block.
+
+### Library Support
+
+Libraries for BLS signature creation are limited in number, although active 
+development appears to be ongoing. Cryptographic algorithms are difficult to
+implement correctly and correctness issues are extremely serious and dangerous.
+No further exploration of BLS should be undertaken without strong assurance of
+a well-tested library with continuing support for creating and verifying BLS
+signatures.
+
+At the moment, there is one candidate, `blst`, that appears to be the most
+mature and well vetted. While this library is undergoing continuing auditing
+and is supported by funds from the Ethereum foundation, adopting a new cryptographic
+library presents some serious risks. Namely, if the support for the library were
+to be discontinued, Tendermint may become saddle with the requirement of supporting
+a very complex piece of software or force a massive ecosystem-wide migration away
+from BLS signatures.
+
+#### Go Standard Library
+
+The Go Standard library has no implementation of BLS signatures.
+
+#### BLST
+
+[blst][blst], or 'blast' is an implementation of BLS signatures written in C
+that provides bindings into Go as part of the repository. This library is
+actively undergoing formal verification by Galois and previously received an
+initial audit by NCC group, a firm I'd never heard of.
+
+`blst` is [targeted for use in prysm][prysm-blst], the golang implementation of Ethereum 2.0.
+
+#### Gnark-Crypto
+
+[Gnark-Crypto][gnark] is a Go-native implementation of elliptic-curve pairing-based
+cryptography. It is completely not auditing and listed as an 'as-is' although
+development appears to be active so formal verification may be forthcoming.
+
+## Open Questions
+
+* *Q*: Can you aggregate a signature twice and still verify?
+* *Q*: If so, how much cost may be imposed by re-including signatures into the same
+* *Q*: Can you aggregate Ed25519 signatures in Tendermint?
```

Comment:
Not just in Tendermint—is there an (any) aggregation scheme for ed25519?

Reply:
In thesis, aggregation is possible for some families of elliptic curves, and ed.... is based on this technology. I think the point here, one of the challenges, it is to find "an" elliptic curve that allows for efficient aggregation while does not provide very poor performance for regular (single signature) singing and verification.

Reply:
Yeah there is one that works, you save ~50% on signature size for a batch of signatures. So a factor of 2 space savings, and some linear factor savings in computation. #7892

Reply:
added reference to the doc.